src/bidiMapper/domains/script/channelProxy.ts

```diff
@@ -183,39 +182,130 @@ export class ChannelProxy {
           ),
           arguments: [
             {
-              objectId: this.#channelHandle,
+              objectId: channelHandle,
             },
           ],
           awaitPromise: true,
-          executionContextId: this.#realm.executionContextId,
+          executionContextId: realm.executionContextId,
           serializationOptions: {
             serialization: 'deep',
-            ...(this.#channel.serializationOptions?.maxObjectDepth ===
+            ...(this.#properties.serializationOptions?.maxObjectDepth ===
               undefined ||
-            this.#channel.serializationOptions.maxObjectDepth === null
+            this.#properties.serializationOptions.maxObjectDepth === null
               ? {}
-              : {maxDepth: this.#channel.serializationOptions.maxObjectDepth}),
+              : {
+                  maxDepth:
+                    this.#properties.serializationOptions.maxObjectDepth,
+                }),
           },
         }
       );
 
-      this.#eventManager.registerEvent(
+      if (message.exceptionDetails) {
+        // TODO: add logging.
+        // TODO: check if a error should be thrown.
+        return;
+      }
+
+      eventManager.registerEvent(
         {
           method: Script.EventNames.MessageEvent,
           params: {
-            channel: this.#channel.channel,
-            data: this.#realm.cdpToBidiValue(
+            channel: this.#properties.channel,
+            data: realm.cdpToBidiValue(
               message,
-              this.#channel.ownership ?? 'none'
+              this.#properties.ownership ?? 'none'
             ),
             source: {
-              realm: this.#realm.realmId,
-              context: this.#realm.browsingContextId,
+              realm: realm.realmId,
+              context: realm.browsingContextId,
             },
           },
         },
-        this.#realm.browsingContextId
+        realm.browsingContextId
       );
     }
   }
+
+  /**
+   * Returns a handle of ChannelProxy from window's property which was set there
+   * by `getEvalInWindowStr`. If window property is not set yet, sets a promise
+   * resolver to the window property, so that `getEvalInWindowStr` can resolve
+   * the promise later on with the channel.
+   * This is needed because `getEvalInWindowStr` can be called before or
+   * after this method.
+   */
+  async #getHandleFromWindow(realm: Realm) {
+    const channelHandleResult = await realm.cdpClient.sendCommand(
+      'Runtime.callFunctionOn',
+      {
+        functionDeclaration: String((id: string) => {
+          // Needed for type script.
+          const w = window as unknown as {
```

Comment:
You can remove this comment, I'd say it's redundant

Reply:
done